test_util/aws.py

```diff
@@ -142,32 +216,22 @@ def get_parameter(self, param):
         raise KeyError('Key not found in template parameters: {}. Parameters: {}'.
                        format(param, self.stack.parameters))
 
-    @retry_boto_rate_limits
-    def get_auto_scaling_instances(self, logical_id):
-        """ Get instances in ASG with logical_id. If logical_id is None, all ASGs will be used
-        Will return instance objects as described here:
-        http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#instance
-
-        Note: there is no ASG resource hence the need for this method
-        """
-        ec2 = self.boto_wrapper.resource('ec2')
-        return [ec2.Instance(i['InstanceId']) for asg in self.boto_wrapper.client('autoscaling').
-                describe_auto_scaling_groups(
-                    AutoScalingGroupNames=[self.stack.Resource(logical_id).physical_resource_id])
-                ['AutoScalingGroups'] for i in asg['Instances']]
-
-    def get_hosts_cached(self, group_name, refresh=False):
-        if refresh or group_name not in self._host_cache:
-            host_list = instances_to_hosts(self.get_auto_scaling_instances(group_name))
-            self._host_cache[group_name] = host_list
-            return host_list
-        return self._host_cache[group_name]
+    def _cached(self, key, value_fn, refresh=False):
```

Comment:
Actually, what I decided when I slept on this is that this whole caching thing needs to go. It makes the code rather hard to maintain and follow, it has an ugly API, it messes with inheritance, and doesnt even work well for the zen stack case. Furthermore, I looked in the code and no where does anyone call these methods in a successive manner that would actually warrant caching this data silently.

Thus, its un-used code and it can go. Its super easy for the caller of this code to implement its own caching. If the caller for whatever reason cannot do this, then we should address this with a more comprehensive approach when that time comes